### DISCUSSION TITLE
等宽字体问题 - Github

### DISCUSSION THREAD

**Discussion by yuuuxt** [F9y4ng/GreasyFork-Scripts#102] [问答专区(Question / Answer)] [answered]
复现: 

- 打开github中任意文件, 如本项目的license: https://github.com/F9y4ng/GreasyFork-Scripts/blob/master/LICENSE (代码应该使用等宽字体)
- 比较`排除渲染github.com`和`重新渲染github.com`的结果(如下, 第一张排除渲染, 第二张重新渲染)

![image](https://user-images.githubusercontent.com/3153534/177003895-6122911b-3111-43b6-881a-e61d37658c2a.png)
![image](https://user-images.githubusercontent.com/3153534/177003903-e0dbeabe-0d39-4998-a65e-2455122e139d.png)

浏览器: firefox 102.0
脚本: 2022.07.02.1

**Comment by F9y4ng** [comment]
问题同 [#101](https://github.com/F9y4ng/GreasyFork-Scripts/discussions/101#discussioncomment-3070030)

**Reply by yuuuxt** [reply]
> 问题同 [#101](https://github.com/F9y4ng/GreasyFork-Scripts/discussions/101#discussioncomment-3070030)

刚测试了一下, 同样是脚本生效, 使用了脚本定义的字体, 而非默认等宽字体 (浏览器: firefox, 同#101)

因此是与#101 不同的问题: #101 是选择生效等宽字体的问题, 这里是本应是等宽字体的被apply非等宽字体的问题

**Reply by F9y4ng** [reply]
## 原因
1. github的代码并不是使用`pre,code`来包含的，而是`table->td`，所以排除渲染操作是无效的。
2. 由于github的等宽字体样式设定优先级比较低，你需要在 **需要渲染的网页元素** 中，追加 `:not([class~='blob-code'])` 保存为站点独享。

**Reply by yuuuxt** [reply]
> ## 原因
> 
>     1. github的代码并不是使用`pre,code`来包含的，而是`table->td`，所以排除渲染操作是无效的。
> 
>     2. 由于github的等宽字体样式设定优先级比较低，你需要在 **需要渲染的网页元素** 中，追加 `:not([class~='blob-code'])` 保存为站点独享。

测试成功, 变成了默认的等宽字体 (但不是脚本设置的等宽字体 - 理想情况), 但字距减小了, 也有混杂非等宽字体(链接处), 对比如下:
关闭脚本: 
![image](https://user-images.githubusercontent.com/3153534/177009409-b1dcf9b3-f86b-4dd2-9820-5094fcc3e251.png)
打开脚本:
![image](https://user-images.githubusercontent.com/3153534/177009416-089f56dd-9464-46eb-8b49-85e7e6948143.png)

**Reply by F9y4ng** [reply]
不建议通过脚本设置英文字体来进行二次渲染，因为字体渲染的样式参数我是以中文为准进行设定的。

纯英文站点，建议你这样用，关闭字体重写功能，仅通过浏览器定义英文等宽字体。

**Reply by yuuuxt** [reply]
> 不建议通过脚本设置英文字体来进行二次渲染，因为字体渲染的样式参数我是以中文为准进行设定的。
> 
> 纯英文站点，建议你这样用，关闭字体重写功能，仅通过浏览器定义英文等宽字体。

其实github也算不上纯英文站点, 中文比例还是有的; 以及我还是有自定义github非等宽字体的需求的, 关闭字体重写就同时舍弃这部分了;

我原来的策略是直接firefix强制指定字体(不勾选允许网页使用其指定字体), 但是这样会影响iconfont.

**Reply by F9y4ng** [reply]
我本地无法复现你的问题。

![屏幕截图 2022-07-03 010757](https://user-images.githubusercontent.com/15855068/177010080-7d9ac833-155d-4201-b696-843d4b240ee3.png)

**Reply by yuuuxt** [reply]
发现关掉[refined github](https://github.com/refined-github/refined-github)就一致了. 
考虑对github的影响, 就设定字体而言, 我觉得还是浏览器强制指定字体更好用点(github也没有icon font); 看起来脚本是无法处理github (+refined github)的复杂情况了?

**Reply by F9y4ng** [reply]
不排除有样式冲突的可能，如果有时间的话，我会看看能否修复这个兼容问题。

**Comment by F9y4ng** [comment, marked as answer]
对于非`pre,code`标签的代码块将不会进行额外筛选，大概分析了几个代码类网站，就打算放弃了，因为工作量实在太大。

针对 #101 所涉及到的自定义等宽字体设置功能，会在下个周更新中增加对应的工具。
